File: 2. Mathematics/ModuloArithmetic.cpp
#include <iostream>
using namespace std;

long long modPower(long long a, long long b, int mod) {
    long long res = 1;
    while (b > 0) {
        if ((b & 1) == 1) {
            res = (res * a) % mod;
        }
        a = (a * a) % mod;
        b = b >> 1;
    }
    return res;
}

int main() {
    long long a = 3978432;
    long long b = 5;
    long long mod = 1000000007;
    long long answer = modPower(a, b, mod);
    cout << answer << endl;
    return 0;
}

File: 2. Mathematics/ModuloArithmetic.js
const mod = 1000000007;

function superPower(x, n, mod) {
  if (n < 0) {
    x = 1 / x;
    n = n * -1;
  }
  if (n == 0) return 1;
  if (x == 0) return 0;

  let res = 1;
  while (n > 0) {
    if (n % 2 == 1) {
      res = (res * x) % mod;
    }
    x = (x * x) % mod;
    n >>= 1;
  }

  return res;
}

console.log(superPower(3978432, 5, mod));
